### DISCUSSION TITLE
What is the format of the `query.expected` files?

### DISCUSSION THREAD

**Discussion by gagliardetto** [github/securitylab#357] [Q&A] [answered]
I'm writing a parser for the `.expected` files (I will then use the data to visually represent the query results).

- Is there some documentation on the format of the `.expected` files?

- Anything in particular I'd need to know when parsing them?

- Also, is the `.expected` file the best source for getting the edges/nodes data?

Thanks!

**Comment by smowton** [comment, marked as answer]
To quote @aibaars from Slack:

It's just plain text. Usually a table-like format like `| value1 | value2 | ... |` If you use named query predicates (query predicate fooBar) then each "table" is prefixed with a "title" like
```
fooBar
| value1 | value2 | ... |
```
The expected output may also contain compiler output such as errors and warnings in addition to the table output.

The above is the case for most tests. However, some query `@kind`s have different output. For example `@kind` graph or `@kind` tree produce some textual graph representation instead of a table.

(Paraphrasing now), for most purposes it is better to run a query with a more firmly-defined format like CSV, or for graph queries DOT or DGML, than to try to parse the `.expected` format that may change from version to version.

**Reply by gagliardetto** [reply]
Thanks!
